packages/components/nodes/vectorstores/Postgres/driver/TypeORM.ts

```diff
@@ -0,0 +1,154 @@
+        const tableName = _tableName ? _tableName : 'documents'
+
+        // Rewrite the method to use pg pool connection instead of the default connection
+        /* Otherwise a connection error is displayed when the chain tries to execute the function
+            [chain/start] [1:chain:ConversationalRetrievalQAChain] Entering Chain run with input: { "question": "what the document is about", "chat_history": [] }
+            [retriever/start] [1:chain:ConversationalRetrievalQAChain > 2:retriever:VectorStoreRetriever] Entering Retriever run with input: { "query": "what the document is about" }
+            [ERROR]: uncaughtException:  Illegal invocation TypeError: Illegal invocation at Socket.ref (node:net:1524:18) at Connection.ref (.../node_modules/pg/lib/connection.js:183:17) at Client.ref (.../node_modules/pg/lib/client.js:591:21) at BoundPool._pulseQueue (/node_modules/pg-pool/index.js:148:28) at .../node_modules/pg-pool/index.js:184:37 at process.processTicksAndRejections (node:internal/process/task_queues:77:11)
+        */
+        instance.similaritySearchVectorWithScore = async (query: number[], k: number, filter?: any) => {
+            return await TypeORMDriver.similaritySearchVectorWithScore(
+                query,
+                k,
+                tableName,
+                await this.getPostgresConnectionOptions(),
+                filter ?? metadataFilters
+            )
+        }
+
+        instance.delete = async (params: { ids: string[] }): Promise<void> => {
+            const { ids } = params
+
+            if (ids?.length) {
+                try {
+                    instance.appDataSource.getRepository(instance.documentEntity).delete(ids)
+                } catch (e) {
+                    console.error('Failed to delete')
+                }
+            }
+        }
+
+        const baseAddVectorsFn = instance.addVectors.bind(instance)
+
+        instance.addVectors = async (vectors, documents) => {
+            return baseAddVectorsFn(vectors, this.sanitizeDocuments(documents))
+        }
+
+        return instance
+    }
+
+    static similaritySearchVectorWithScore = async (
+        query: number[],
+        k: number,
+        tableName: string,
+        postgresConnectionOptions: ICommonObject,
+        filter?: any
+    ) => {
+        const embeddingString = `[${query.join(',')}]`
+        let _filter = '{}'
+        let notExists = ''
+        if (filter && typeof filter === 'object') {
+            if (filter.$notexists) {
+                notExists = `OR NOT (metadata ? '${filter.$notexists}')`
+                delete filter.$notexists
+            }
+            _filter = JSON.stringify(filter)
+        }
+
+        const queryString = `
+            SELECT *, embedding <=> $1 as "_distance"
+            FROM ${tableName}
+            WHERE metadata @> $2
+            ${notExists}
+            ORDER BY "_distance" ASC
+            LIMIT $3;`
```

Comment:
I'm dibutative about this logic... (i paste it as is from legacy code)

The WHERE clause sounds like: Where `metadata` includes `filter` OR `metadata` keys not includes `chatflowIdKey`
which seems to make the first part optional when second part is truthy, and this is clearly not the expected behavior

Reply:
In PGVector driver, i removed the chatflow part from filters and added a ` AND (<compareExistingValue> OR <keyNotExist>)`

Reply:
The original intention is to filter these conditions:
- metadata contains the JSON structure (filter)
OR
- metadata does not contains the key `flowise_chatId`

Say the following scenario:
1.) Owner of the chatflow upserted data to the postgres vector store through API, ex: `[{pageContent: 'abc', metadata: {}}]`
2.) User A interacts with the chatflow and upload some files. This will be upserted on the fly, resulting in `[{pageContent: 'this is userA', metadata: { flowise_chatId: userA-session }}]`
3.) User B interacts with the chatflow and upload some files. This will be upserted on the fly, resulting in `[{pageContent: 'this is userB', metadata: { flowise_chatId: userB-session }}]`

Now, when user A is asking questions, we want to only get 2 sources:
```
[
  {pageContent: 'abc', metadata: {}}
  {pageContent: 'this is userA', metadata: { flowise_chatId: userA-session }}
]
```
We do not want to get source with metadata of `flowise_chatId: userB-session`

Hence this results in the following filters:
metadata contains that specific chatId OR metadata that doesn't have the `flowise_chatId` key

Reply:
Tanks for your response @HenryHengZJ 🙏 

We're ok about the expected behavior 👍 

Now imagine a 4th scenario:
For whatever reason, i add a metadata filter on the vector store node, ex : `[{pageContent: 'this is userA', metadata: { flowise_chatId: userA-session, foo: 'bar' }}]`

With the actual or clause, the or clause will be: 

```sql
WHERE metadata @> { flowise_chatId: userA-session, foo: 'bar' }
OR NOT (metadata ? 'flowise_chatId')
```

This will return:
- Chatflow uploaded documents filtered with `foo=bar` 
- Any documents without `flowise_chatId`, not filtered with `foo=bar` 

If we want the right behavior, the where clause should be:

```sql
WHERE metadata @> { foo: 'bar' }
AND (metadata @> { flowise_chatId: userA-session } OR  NOT (metadata ? 'flowise_chatId'))
```

Reply:
as an alternative, we can also do:

```sql
WHERE metadata @> '{ foo: \'bar\' }'
OR metadata @> '{ flowise_chatId: \'userA-session\' }'
```

Which means:
- Any document with `foo=bar`
- Chatflow uploaded documents

Reply:
Long story short, you are absolutely right, we definitely want the 2nd query:
```sql
WHERE metadata @> '{ foo: \'bar\' }'
AND (metadata @> '{ flowise_chatId: \'userA-session\' }' OR NOT (metadata ? 'flowise_chatId'))
```

Would you be able to make the changes? for both typeorm and pgvector

Reply:
of course

Reply:
Done in https://github.com/FlowiseAI/Flowise/pull/3367/commits/2627b766ac32230f35066002af68b3ee873a08f8

Reply:
@HenryHengZJ After more reflection, i think this query will match all the cases:

```sql
WHERE (metadata @> '{ foo: \'bar\' }' AND NOT (metadata ? 'flowise_chatId'))
OR metadata @> '{ flowise_chatId: \'userA-session\' }'
```

Reply:
Hmm I think its 3 clauses:

- Any documents with metadata contains `foo: bar`
- or any documents with metadata contains `foo: bar`, AND `flowise_chatId: <chatid>`
- or any documents with metadata contains `foo: bar`, AND without `flowise_chatId`

Reply:
IMO we don't care about `foo: bar` until there is a `flowise_chatId` set, because if it's been uploaded via the chatflow, it's always part of the chatflow context (and filters will not be added to the document's metadata).

Basically, the expectations are:
- Return any document uploaded via the chatflow
- Return filtered documents from the vector store
- Don't return the documents uploaded via other chatflows

Reply:
The part:
```sql
AND NOT (metadata ? 'flowise_chatId')
```
is just present to ensure that we don't return documents from other chatflows if there is no filter set:

```sql
WHERE (metadata @> '{}' AND NOT (metadata ? 'flowise_chatId'))
OR metadata @> '{ flowise_chatId: \'userA-session\' }'
```

Reply:
Okay so this will results if either of these two conditions is true:
- The metadata contains `foo': 'bar` and does not have `flowise_chatId`.
- The metadata contains `flowise_chatId': 'userA-session`, regardless of whether `foo': 'bar` is present or not.

Rewind back the scenario, and see if this fits:
1.) Owner upserted documents without metadata: `pageContent: 'this is doc 1', metadata: {}`
2.) Owner upserted documents WITH metadata: `pageContent: 'this is doc 2', metadata: { source: doc2 }`
Owner publish the chatflow to public
3.) User A uploaded a file from the chat window, results in `pageContent: 'this is user A', metadata: { flowise_chatId: userA }`
4.) User B uploaded a file from the chat window, results in `pageContent: 'this is user B', metadata: { flowise_chatId: userB }`

Now 4 docs are sitting in vector store:
1.) `pageContent: 'this is doc 1', metadata: {}`
2.) `pageContent: 'this is doc 2', metadata: { source: doc2 }`
3.) `pageContent: 'this is user A', metadata: { flowise_chatId: userA }`
4.) `pageContent: 'this is user B', metadata: { flowise_chatId: userB }`

Scenario 1: Owner doesnt set any filter on metadata. User A when chatting, should only get documents (1) (2) and (3)

Scenario 2: Owner pre set filter on metadata. User A when chatting, should only get documents (2) and (3)

Looks like the new clause you have there satisfy these 2 scenarios

Reply:
Fixed in https://github.com/FlowiseAI/Flowise/pull/3367/commits/0a39892ec56a34745285b3b66ec26004816a5ee5